source/common/conn_pool/conn_pool_base.cc

```diff
@@ -33,11 +33,31 @@ void ConnPoolImplBase::destructAllConnections() {
   dispatcher_.clearDeferredDeleteList();
 }
 
-void ConnPoolImplBase::tryCreateNewConnection() {
-  if (pending_requests_.size() <= connecting_request_capacity_) {
-    // There are already enough CONNECTING connections for the number
-    // of queued requests.
-    return;
+bool ConnPoolImplBase::shouldCreateNewConnection() const {
+  // The number of requests we want to be provisioned for is the number of
+  // pending and active requests times the prefetch ratio.
+  // The number of requests we are (theoretically) provisioned for is the
+  // connecting request capacity plus the number of active requests.
+  //
+  // If prefetch ratio is not set, it defaults to 1, and this simplifies to the
+  // legacy value of pending_requests_.size() > connecting_request_capacity_
+  return (pending_requests_.size() + num_active_requests_) * host_->cluster().prefetchRatio() >
+         (connecting_request_capacity_ + num_active_requests_);
+}
+
+void ConnPoolImplBase::tryCreateNewConnections() {
+  // Prefetch ratio is capped at 3, so we should never need to fetch more than 3 at a time.
```

Comment:
can you assert this invariant somehow? Also, even though this is config driven do you want a kill switch? Might be worth it.

Reply:
Right now it's an invariant, and I could add asserts, but we'll lose the invariant as soon as we stop prefetching for unhealthy hosts, at which point when a host becomes healthy we could have an arbitrarily large number of needed connections.

I'll instead say it's roughly arbitrary, and should not occur in the common case unless hosts go unhealthy.